consensus/hotstuff/committees/consensus_committee.go

```diff
@@ -11,35 +11,83 @@ import (
+// outside all cached epochs. This can happen when a query is made for a view in the
+// next epoch, if that epoch is not committed yet. This can also happen when an
+// old epoch is queried (>3 in the past), even if that epoch does exist in storage.
+var ErrViewForUnknownEpoch = fmt.Errorf("by-view query for unknown epoch")
+
+// staticEpochInfo contains leader selection and the initial committee for one epoch.
+type staticEpochInfo struct {
+	firstView        uint64
+	finalView        uint64
+	leaders          *leader.LeaderSelection
+	initialCommittee flow.IdentityList
+	dkg              hotstuff.DKG
+}
+
+// newStaticEpochInfo returns the static epoch information from the epoch.
+// This can be cached and used for all by-view queries for this epoch.
+func newStaticEpochInfo(epoch protocol.Epoch) (*staticEpochInfo, error) {
+	firstView, err := epoch.FirstView()
+	if err != nil {
+		return nil, fmt.Errorf("could not get first view: %w", err)
+	}
+	finalView, err := epoch.FinalView()
+	if err != nil {
+		return nil, fmt.Errorf("could not get final view: %w", err)
+	}
+	leaders, err := leader.SelectionForConsensus(epoch)
+	if err != nil {
+		return nil, fmt.Errorf("could not get leader selection: %w", err)
+	}
+	initialidentities, err := epoch.InitialIdentities()
+	if err != nil {
+		return nil, fmt.Errorf("could not initial identities: %w", err)
+	}
+	initialCommittee := initialidentities.Filter(filter.IsVotingConsensusCommitteeMember)
+	dkg, err := epoch.DKG()
+	if err != nil {
+		return nil, fmt.Errorf("could not get dkg: %w", err)
+	}
+
+	epochInfo := &staticEpochInfo{
+		firstView:        firstView,
+		finalView:        finalView,
+		leaders:          leaders,
+		initialCommittee: initialCommittee,
+		dkg:              dkg,
+	}
+	return epochInfo, nil
+}
 
 // Consensus represents the main committee for consensus nodes. The consensus
 // committee persists across epochs.
 type Consensus struct {
-	mu      sync.RWMutex
-	state   protocol.State                     // the protocol state
-	me      flow.Identifier                    // the node ID of this node
-	leaders map[uint64]*leader.LeaderSelection // pre-computed leader selection for each epoch
+	mu     sync.RWMutex
+	state  protocol.State              // the protocol state
+	me     flow.Identifier             // the node ID of this node
+	epochs map[uint64]*staticEpochInfo // cache of initial committee & leader selection per epoch
 }
 
-var _ hotstuff.Committee = (*Consensus)(nil)
+var _ hotstuff.Replicas = (*Consensus)(nil)
+var _ hotstuff.DynamicCommittee = (*Consensus)(nil)
 
 func NewConsensusCommittee(state protocol.State, me flow.Identifier) (*Consensus, error) {
 
 	com := &Consensus{
-		state:   state,
-		me:      me,
-		leaders: make(map[uint64]*leader.LeaderSelection),
+		state:  state,
+		me:     me,
+		epochs: make(map[uint64]*staticEpochInfo),
 	}
 
 	final := state.Final()
 
 	// pre-compute leader selection for current epoch
 	current := final.Epochs().Current()
```

Comment:
As best practise, I prefer to to only query final once, because the return value can theoretically change between different queries. We query twice in this method: [here](https://github.com/onflow/flow-go/blob/4de4231e43a65fe672a54aab4f84971ecac03db6/consensus/hotstuff/committees/consensus_committee.go#L89) and [here](https://github.com/onflow/flow-go/blob/4de4231e43a65fe672a54aab4f84971ecac03db6/consensus/hotstuff/committees/consensus_committee.go#L101)

I don't think there is an issue with the way it is currently implemented, because finality can probably not progress within the constructor. Nevertheless, I think it is no extra hassle to stick to our best practise and write
```golang
	eq := final.Epochs()
	current := eq.Current()
        ...
	previous := eq.Previous()
```

Reply:
We query a particular snapshot only once though, on line 86 (`final := state.Final()`). Once we capture the snapshot once, querying it multiple times afterward is safe.